### ISSUE TITLE
[bug] conan test fails when using lockfiles

### ISSUE THREAD

**Issue by DukeOfKirkcaldy** [conan-io/conan#10916]
The test_package tests fail for both `conan create` or `conan test` commands when using lockfiles.

### Environment Details (include every applicable attribute)
  * Conan version: 1.46.0
  * Python version: 3.7.9

### Steps to reproduce (Include if Applicable)
```
conan lock create conanfile.py --profile:host default --profile:build default --lockfile-out locks/conan.lock
conan create conanfile.py --lockfile locks/conan.lock
```

### Logs (Executed commands with output) (Include/Attach if Applicable)
```
<install and build succeeds>
...
Executing test_package liba/0.1
Configuration (profile_host):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=7
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

Configuration (profile_build):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=7
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

ERROR: Something failed while testing 'liba/0.1' test_package after it was built using the lockfile. Please report this error: 'liba'
Error: Process completed with exit code 1.
```

More detailed log from `cd test_package && conan test . liba/0.1 --lockfile=../locks/conan.lock` (also tried with reference `liba/0.1@_/_`):
```
Using lockfile: 'C:\Projects\conan-demo\conan-demo-liba\conan.lock'
Configuration (profile_host):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=7
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

Configuration (profile_build):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=7
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

Traceback (most recent call last):
  File "c:\users\gjess\appdata\local\programs\python\python37\lib\site-packages\conans\client\command.py", line 2238, in run
    method(args[0][1:])
  File "c:\users\gjess\appdata\local\programs\python\python37\lib\site-packages\conans\client\command.py", line 311, in test
    lockfile=args.lockfile, profile_build=profile_build)
  File "c:\users\gjess\appdata\local\programs\python\python37\lib\site-packages\conans\client\conan_api.py", line 93, in wrapper
    return f(api, *args, **kwargs)
  File "c:\users\gjess\appdata\local\programs\python\python37\lib\site-packages\conans\client\conan_api.py", line 334, in test
    test_build_folder=test_build_folder, recorder=recorder)
  File "c:\users\gjess\appdata\local\programs\python\python37\lib\site-packages\conans\client\cmd\test.py", line 51, in install_build_and_test
    test=True  # To keep legacy test_package_layout
  File "c:\users\gjess\appdata\local\programs\python\python37\lib\site-packages\conans\client\manager.py", line 62, in deps_install
    require_overrides=require_overrides)
  File "c:\users\gjess\appdata\local\programs\python\python37\lib\site-packages\conans\client\graph\graph_manager.py", line 124, in load_graph
    require_overrides)
  File "c:\users\gjess\appdata\local\programs\python\python37\lib\site-packages\conans\client\graph\graph_manager.py", line 164, in _load_root_node
    require_overrides)
  File "c:\users\gjess\appdata\local\programs\python\python37\lib\site-packages\conans\client\graph\graph_manager.py", line 281, in _load_root_test_package
    graph_lock.find_require_and_lock(create_reference, conanfile)
  File "c:\users\gjess\appdata\local\programs\python\python37\lib\site-packages\conans\model\graph_lock.py", line 664, in find_require_and_lock
    conanfile.requires[reference.name].lock(locked_ref, node_id)
KeyError: 'liba'

ERROR: 'liba'
```

**Comment by edwardvandevorst-philips** [comment]
I have exactly the same problem. How to work around this?

**Comment by pwueje** [comment]
Is there any news to this issue?

We face the same issue that prevents us from using lock files. I've created a sample example that illustrates the error: https://github.com/koenig-bauer-coding/Conan-Lock-File-Error

If this bug isn't going to be resolved within conan 1.x a quick info would be nice, so we could try to find a workaround for us.

**Comment by LevWi** [comment]
I have same problem if test_packages directories used  for conan v2 
If you use  conan-center-index then check test_v1_package directories. It's help me

**Comment by cassava** [comment]
I'm also having this problem. @LevWi how did the `test_v1_package` directories help you?

**Comment by LevWi** [comment]
@cassava 
For example. Instead of 
conan test ./recipes/libjpeg-turbo/all/test_package libjpeg-turbo/2.1.2@
I used
conan test ./recipes/libjpeg-turbo/all/test_v1_package libjpeg-turbo/2.1.2@

As I can see  test_package and  test_v1_package have different conanfile.py structure

**Comment by pwueje** [comment]
@LevWi thanks for your hint, but if I test your workaround, it seems that the package test won't be executed at all for `conan create`.

I guess, you execute the test separate afterwards?

**Comment by LevWi** [comment]
@pwueje 
Yes. I use tests after creating packages.
You can try this way
conan create -tf <test_v1_package> ...
https://docs.conan.io/en/latest/reference/commands/creator/create.html

**Comment by pwueje** [comment]
If I use `conan create` like suggested, I'm running into the same error as just executing it without the `tf` parameter und the default test_package folder.

@franramirez688 are there any chances that this behaviour will be fixed or do we have to live with it until the big conan 2.x bang?

**Comment by edwardvandevorst-philips** [comment]
I have another workaround (I am not 100% sure it is always correct).

1. First create lock files
2. Export the package using the lock files (`conan export`)
3. Now the recipe is in the local cache. After this I run `conan create` without lock files and explicity update assuming it will use the exported recipe in the cache. Also the test is done in this step.

The above happens on a build machine with a clean cache.

**Comment by franramirez688** [comment]
> @franramirez688 are there any chances that this behaviour will be fixed or do we have to live with it until the big conan 2.x bang?

@pwueje Let me discuss this with the team. I tend to say that it was not easy to solve this for 1.x, but I'm not 100% sure.

**Comment by ds-frankgi** [comment]
> 3\. After this I run `conan create` without lock files

@edwardvandevorst-philips 
Not sure if I understand you correctly but if you leave out the "--lockfile locks/conan.lock" part in that call the prior effort to "conan lock create" makes no sense to me. It is essential to have this here. 
Idea is to create a package with a lockfile to keep the same dependencies (e.g. in CI) and/or creating buildinfo to scan for vulnerabilities, imho.
So, "conan create" will of course run in this case (unless there are no other issues).

**Comment by edwardvandevorst-philips** [comment]
My assumption is that when the recipe is in the local cache and you don't use `-u` when doing `conan create` it uses the information from the recipe and so the dependencies that are listed in that exported recipe. But I could be wrong here.

**Comment by franramirez688** [comment]
Hi @pwueje,

After discussing it with the team, I'm afraid we cannot easily solve it for Conan 1.x. The complexity of this change is important, and hopefully, it won't happen for Conan 2.x anymore (see https://github.com/conan-io/conan/issues/10966). Sorry for any inconvenience.

**Comment by pwueje** [comment]
Hi @franramirez688,

thanks for your investigation. It's a shame, but therefore we're much more looking forward to 2.x. Thanks for the great work the Conan team is doing so far.

**Comment by ds-frankgi** [comment]
> My assumption is that when the recipe is in the local cache and you don't use `-u` when doing `conan create` it uses the information from the recipe and so the dependencies that are listed in that exported recipe. But I could be wrong here.

@edwardvandevorst-philips 
I tried this and it looked like it worked for linux builds but it failed for windows builds in my environment. I found that conan export is using a default configuration for exporting the package DIFFERENT from that placed in the lockfile (for conan 1.53). Even if I build a linux configuration e.g. with gcc conan exports writes config for clang. For linux the subsequent build via conan create continues somehow but for windows it breaks as we do not compile in windows with clang.
So unfortunately this is not the final solution for us.

@franramirez688: 
Is it ensured that this bug is solved in conan2.x with all possible arguments given?
The link [#10966] you posted just contained a misconfigured command.
We also provide arguments like "build_type={type}", "-o:b=...", f"-o:h=...", "-c:h=tools.cmake.cmaketoolchain:generator={a.generator}" which should find their way into the lockfile.
